ksqldb-engine/src/main/java/io/confluent/ksql/planner/LogicalPlanner.java

```diff
@@ -575,7 +579,107 @@ private JoinNode buildJoin(final Join root, final String prefix, final boolean i
     );
   }
 
+  private void verifyJoin(final Join root, final PlanNode left, final PlanNode right) {
+    if (left.getNodeOutputType() == DataSourceType.KSTREAM) {
+      if (right.getNodeOutputType() == DataSourceType.KTABLE) {
+
+        // stream-table join detected
+
+        if (root.getInfo().getType().equals(JoinType.OUTER)) {
+          throw new KsqlException("Invalid join type:"
+              + " full-outer join not supported for stream-table joins.");
+        }
+
+        // added outer check to make `multi-joins.json` pass
+        // the test uses expressions -> need to follow up on this issue
+        if (root.getInfo().getRightJoinExpression() instanceof ColumnReferenceExp) {
+          if (joinOnNonKeyAttribute(
+              ((ColumnReferenceExp) root.getInfo().getRightJoinExpression()),
+              right
+          )) {
+            throw new KsqlException("Invalid join condition:"
+                + " stream-table joins require to join on the table's primary key.");
+          }
+        }
+      }
+
+      // stream-stream join detected: nothing to verify
+
+    } else {
+      if (right.getNodeOutputType() == DataSourceType.KSTREAM) {
+        throw new KsqlException("Invalid join order:"
+            + " table-stream joins are not supported; only stream-table joins.");
+      }
+
+      // table-table join detected
+
+      if (joinOnNonKeyAttribute(
+            ((ColumnReferenceExp) root.getInfo().getLeftJoinExpression()),
+            left
+      )) {
+        throw new KsqlException("Invalid join condition:"
+            + " table-table joins require to join on the primary key of the left input table."
+        );
+      }
+
+      if (joinOnNonKeyAttribute(
+            ((ColumnReferenceExp) root.getInfo().getRightJoinExpression()),
+            right
+      )) {
+        throw new KsqlException("Invalid join condition:"
+            + " foreign-key table-table joins are not supported.");
+      }
+    }
+  }
+
+  private boolean joinOnNonKeyAttribute(final ColumnReferenceExp joinExpression,
+                                        final PlanNode node) {
+    final String joinAttributeName = joinExpression.getColumnName().text();
+    final List<DataSourceNode> dataSourceNodes = node.getSourceNodes().collect(Collectors.toList());
+    final String singleAttributeKeyName;
+    if (dataSourceNodes.size() == 1) {
+      singleAttributeKeyName = dataSourceNodes.get(0).getSchema().key().get(0).name().text();
```

Comment:
Why are we assuming there's a single key? Shouldn't we check whether there are multiple, and return true if so?

Reply:
We don't support joins on structured keys yet, and if one tries to do this, we should not reach this point in the code but fail earlier.

Reply:
Maybe adding a TODO in case we forget to update it when adding multi-key join support.